src/data/sigma/interval.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.sigma.order
+import order.locally_finite
+
+/-!
+# Finite intervals in a sigma type
+
+This file provides the `locally_finite_order` instance for the disjoint sum of orders `Σ i, α i` and
+calculates the cardinality of its finite intervals.
+
+## TODO
+
+Do the same for the lexicographical order
+-/
+
+open finset function
+
+namespace sigma
+variables {ι : Type*} {α : ι → Type*}
+
+/-! ### Disjoint sum of orders -/
+
+section disjoint
+variables [decidable_eq ι] [Π i, preorder (α i)] [Π i, locally_finite_order (α i)]
+
+instance : locally_finite_order (Σ i, α i) :=
+{ finset_Icc := λ a b, dite (a.1 = b.1)
+    (λ h, (Icc (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
+  finset_Ico := λ a b, dite (a.1 = b.1)
+    (λ h, (Ico (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
+  finset_Ioc := λ a b, dite (a.1 = b.1)
+    (λ h, (Ioc (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
+  finset_Ioo := λ a b, dite (a.1 = b.1)
+    (λ h, (Ioo (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
+  finset_mem_Icc := λ ⟨i, a⟩ ⟨j, b⟩ x, begin
```

Comment:
Is there any way to golf this? It feels like you are doing (almost) the same thing four times. If it can't be golfed with a lemma, I would still be tempted to write a throw-away tactic macro to deduplicate the proofs.

Reply:
I can always use `locally_finite_order.of_Icc` or `locally_finite_order.of_Icc'`, but at the expense of greater instance assumptions. I already golfed it quite a lot using `mk_le_mk_iff ` and `mk_lt_mk_iff `. Not sure there's much to gain from writing a tactic.

PS: I don't know tactic-writing (yet!)

Reply:
```
namespace tactic.interactive

meta def finset_mem_helper : tactic unit :=
`[ rintro ⟨i, a⟩ ⟨j, b⟩ x,
    rcases eq_or_ne i j with rfl | hij,
    { rw [dif_pos rfl, mem_map],
      split,
      { rintro ⟨x, hx, rfl⟩,
        simpa using hx },
      { rintro ⟨⟨i, a, x, ha⟩, ⟨_, _, b, hb⟩⟩,
        simp [ha, hb] } },
    { simp only [dif_neg hij, false_iff, not_mem_empty],
      rintro ⟨⟨i, a, x, ha⟩, ⟨_, _, b, hb⟩⟩,
      exact hij rfl } ]

end tactic.interactive

namespace sigma
variables {ι : Type*} {α : ι → Type*}

/-! ### Disjoint sum of orders -/

section disjoint
variables [decidable_eq ι] [Π i, preorder (α i)] [Π i, locally_finite_order (α i)]

instance : locally_finite_order (Σ i, α i) :=
{ finset_Icc := λ a b, dite (a.1 = b.1)
    (λ h, (Icc (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
  finset_Ico := λ a b, dite (a.1 = b.1)
    (λ h, (Ico (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
  finset_Ioc := λ a b, dite (a.1 = b.1)
    (λ h, (Ioc (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
  finset_Ioo := λ a b, dite (a.1 = b.1)
    (λ h, (Ioo (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅),
  finset_mem_Icc := by finset_mem_helper,
  finset_mem_Ico := by finset_mem_helper,
  finset_mem_Ioc := by finset_mem_helper,
  finset_mem_Ioo := by finset_mem_helper }
  ```
  is an option if golfing with a lemma is impossible

Reply:
Here's another option:
```
def sigma_lift (f : ∀ ⦃i⦄, α i → α i → finset (α i)) (a b : sigma α) : finset (sigma α) :=
dite (a.1 = b.1) (λ h, (f (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅)

lemma mem_sigma_lift' (f : ∀ ⦃i⦄, α i → α i → finset (α i)) (a b x : sigma α) :
  x ∈ sigma_lift f a b ↔ ∃ (ha : a.1 = x.1) (hb : b.1 = x.1), x.2 ∈ f (ha.rec a.2) (hb.rec b.2) :=
begin
  obtain ⟨i, a⟩ := a,
  obtain ⟨j, b⟩ := b,
  rcases decidable.eq_or_ne i j with rfl | hij,
  { split,
    { simp only [sigma_lift, embedding.sigma_mk_apply, dif_pos rfl, mem_map],
      rintro ⟨x, hx, rfl⟩,
      exact ⟨rfl, rfl, hx⟩ },
    { rintro ⟨⟨⟩, ⟨⟩, h⟩,
      simp only [sigma_lift, dif_pos rfl, mem_map],
      exact ⟨_, h, by simp [sigma.ext_iff]⟩ } },
  simp only [sigma_lift, dif_neg hij, not_mem_empty, false_iff],
  rintro ⟨⟨⟩, ⟨⟩, _⟩,
  apply hij rfl
end

lemma mk_le_mk_iff' [Π i, has_le (α i)] {i j : ι} {a : α i} {b : α j} :
  (⟨i, a⟩ : sigma α) ≤ ⟨j, b⟩ ↔ ∃ (h : i = j), h.rec a ≤ b :=
⟨by { rintro ⟨a, _, b, h⟩, exact ⟨rfl, h⟩ }, by { rintro ⟨rfl, h⟩, exact le.fiber _ _ _ h }⟩

def sigma_Icc (a b : sigma α) : finset (sigma α) :=
  sigma_lift (λ _, Icc) a b

lemma mem_sigma_Icc : ∀ (a b x : sigma α), x ∈ sigma_Icc a b ↔ a ≤ x ∧ x ≤ b :=
begin
  rintro ⟨_, _⟩ ⟨_, _⟩ ⟨_, _⟩,
  simp_rw [sigma_Icc, mem_sigma_lift', mk_le_mk_iff', mem_Icc, exists_and_distrib_left,
    ←exists_and_distrib_right, ←exists_prop],
  exact bex_congr (λ _ _, by split; rintro ⟨⟨⟩, ht⟩; exact ⟨rfl, ht⟩),
end
```
This way we have `sigma_lift` which can be used for infinite intervals and other fiber-wise constructions, and the proof of each interval is shorter. (Of course `sigma_Icc` shouldn't be a standalone def, and `mem_sigma_Icc` should be inside the instance def)

Reply:
Opened #10958 for this new definition.